### PR TITLE
Shared cache on NFS Introduced

### LINKED ISSUE #103: guide: using NFS as a remote storage
I.e. mount it to some dir and specify it as a local dir repository for dvc.

Related to https://github.com/iterative/dvc.org/issues/4303

### PR DESCRIPTION
**Pull request by ryokugyu**
fix #103

### CODE REVIEW

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+With large data files it is better to set the cache directory to external NFS.
+Not only just it will cache the data faster but also version the data. Suppose,
+we have a dataset with 1 million images. With DVC, we can have multiple versions
+of a dataset without affecting each other work and without creating duplicates
+of a complete dataset. With `cache directory` set to `NFS server` you would
+avoid copying large files from NFS server to the machine and DVC will manage the
+links from the workspace to cache. For more information, visit
+[Data and Model Files Versioning](/doc/use-cases/data-and-model-files-versioning).
+
+### Preparation
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: all these should be ## instead of ### to show them on the right-side menu

Comment (shcheklein) [reply static/docs/use-cases/shared-storage-on-nfs.md]: or almost all of them

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# Shared Storage on NFS
+
+In the modern software development environment, teams are working together on
+same dataset to get the results. It became necessary that data is accessible and
+every team member has a same updated dataset. For this example, we will be using
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: it's better to say that attached NFS storage is used as a shared storage usually ... not "we will ..." ... it's not about us, it's about NFS being widely used for this scenario. Here we are just explaining how DVC can help manage data in this case.

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+NFS (Network File System) for storing and sharing files on the network. This
+allows you to have better resource utilization such as ability to store large
+disk consuming dataset on a single host machine.
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: just large datasets

i'm not sure I understand this sentence, what exactly the single host machine does here?

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+For optimizing the performance, we can set the `cache directory` on NFS server
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: we can setup a shared DVC cache on NFS ... and use it across multiple clients that are mounting this NFS ...

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+by configuring the DVC repository from making changes in the DVC config file
+which is present in `.dvc/config` location. With DVC, you can easily setup a
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: may be actually skip the first sentence

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+store data for your projects effectively as possible and have a workspace
+restoration/switching speed as instant as `git checkout` for your code.
+
+With large data files it is better to set the cache directory to external NFS.
+Not only just it will cache the data faster but also version the data. Suppose,
+we have a dataset with 1 million images. With DVC, we can have multiple versions
+of a dataset without affecting each other work and without creating duplicates
+of a complete dataset. With `cache directory` set to `NFS server` you would
+avoid copying large files from NFS server to the machine and DVC will manage the
+links from the workspace to cache. For more information, visit
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: remove the last sentence. There should be a better link.

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+team member is using the same cache location.
+
+After configuring NFS on both server and client side. Let's create an export
+directory on server side where all data will be stored.
+
+```dvc
+$ mkdir -p /storage
+```
+
+You will have to make sure that the directory has proper permissions setup,so
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: "setup, so" (space is missing)

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+$ mkdir -p /mnt/dataset/
+```
+
+From `/mnt/dataset/` you will be able to access `/storage` directory present in
+host server from your local machine.
+
+## Configuring Cache location
+
+After mounting the shared directory on client side. Assuming project code is
+present in `/home/user/project1`. Let's initialize a `dvc repo`.
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: let's use just `project`

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+```dvc
+$ cd /home/user/project1/
+$ git init
+$ dvc init
+$ git add .dvc .gitignore
+$ git commit . -m "initialize DVC"
+```
+
+With `dvc init`, we initialized a DVC repository. DVC will start tracking all
+the changes.
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: DVC is not tracking anything yet. Let's remove the second sentence. Put read more here https://dvc.org/doc/get-started/initialize

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+Tell DVC to use the directory we've set up as an external cache location by
+running:
+
+```dvc
+$ dvc config cache.dir /mnt/dataset/storage
+$ dvc config cache.type "reflink,symlink,hardlink,copy"
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: split all this commands, put explanation for each of them

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+$ dvc config cache.protected true
+$ git add .dvc .gitignore
+$ git commit . -m "DVC cache location updated"
+```
+
+By default cache is present in the `.dvc/cache` location. `dvc cache dir`
+changes the location of cache directory to `/mnt/dataset/storage`
+
+`config cache.dir /path/to/cache/directory` - sets cache directory location.
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: no need to complicate it here with alternatives, let's use `dvc cache dir`

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+Alternatively, we can also use `dvc cache dir /path/to/cache/directory`.
+
+`cache.type "reflink,symlink,hardlink,copy"` - link type that DVC should use to
+link data files from cache to your workspace. It enables symlinks to avoid
+copying large files.
+
+`cache.protected true` - to make links `read only` so that we you don't corrupt
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: again explain it better that since we are going to use symlinks in this case between cache and workspace (since they are located on different file systems) it important to protect files so that we don't corrupt the cache accidentally. Mention that `dvc unprotect` should be used in this case, link to the https://dvc.org/doc/user-guide/update-tracked-file

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+data accidentally present in the workspace.
+
+For more information on `config` options, visit
+[here](https://dvc.org/doc/commands-reference/config#configuration-sections)
+
+Also, let git know about the changes we have done.
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: Git.

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+## Add data to DVC cache
+
+Now, add first version of the dataset into the DVC cache (this is done once for
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: need some explanation why and when is it needed - clients do not have enough space on their file systems to keep even a single version of a dataset.

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+a dataset).
+
+```dvc
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: let's simplify all this workflow. Let's just ask users to SSH into NFS serve machine, do `git clone .../project`. Move data into `project` and run `dvc add`, `git commit`, `git push`, (`dvc push`  optional) after that. All the stuff below can be adjusted a bit.

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 3 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: @shcheklein we are storing the dataset on a single machine and people are accessing it from there local machines. Not everyone wants to copy the dataset locally on their machine and work.

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 2 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: Done

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 5 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: removed

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 8 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: done

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 9 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: done.

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 11 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: done

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 10 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: done

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+$ git commit . -m "DVC cache location updated"
+```
+
+By default cache is present in the `.dvc/cache` location. `dvc cache dir`
+changes the location of cache directory to `/mnt/dataset/storage`
+
+`config cache.dir /path/to/cache/directory` - sets cache directory location.
+Alternatively, we can also use `dvc cache dir /path/to/cache/directory`.
+
+`cache.type "reflink,symlink,hardlink,copy"` - link type that DVC should use to
```
[the remark this answers is not in the document]
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: link introduced. later, I will try to explain it here in a better manner.

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 12 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: @shcheklein we need `dvc unprotect` only when we are writing to NFS directly? 

As mentioned [here](https://discuss.dvc.org/t/share-nas-data-in-server/180/7), when `cache.protected true` and `cache.protected false`  it doesn't work.

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 15 above)
Comment (ryokugyu) [reply static/docs/use-cases/shared-storage-on-nfs.md]: @shcheklein i think it will just confuse the user.

**Comment by ryokugyu** [comment]
@shcheklein please review this.

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 15 above)
Comment (shcheklein) [reply static/docs/use-cases/shared-storage-on-nfs.md]: `cp -r . /project/` is very confusing also. I would say we need to explain the motivation here - we want to avoid copying existing data to a client machine to take it under DVC control.

I also, think `git clone` protocol is a standard way to collaborate and update different requirements. It's better to do this from the NFS server machine. It'll emphasize that NFS takes care about data.

File: static/docs/use-cases/shared-storage-on-nfs.md (hunk 12 above)
Comment (shcheklein) [reply static/docs/use-cases/shared-storage-on-nfs.md]: we need to run `dvc unprotect` in the client's workspace if we want to edit/rewrite the file that is under DVC control.

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+is widely used for storing and sharing files on the network. This allows you to
+have better resource utilization such as ability to store large datasets on a
+single host machine.
+
+With DVC, you can easily setup a shared cache storage on the NFS server that
+will allow your team to share and store data for your projects effectively as
+possible and have a workspace restoration/switching speed as instant as
+`git checkout` for your code.
+
+With large data files it is better to set the cache directory to external NFS.
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: minor: we use `it's` - it's less formal

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+Not only just it will cache the data faster but also version the data. Suppose,
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: `cache faster` - I'm not sure I understand this

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+we have a dataset with 1 million images. With DVC, we can have multiple versions
+of a dataset without affecting each other work and without creating duplicates
+of a complete dataset. With `cache directory` set to `NFS server` you would
+avoid copying large files from NFS server to the machine and DVC will manage the
+links from the workspace to cache.
+
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: the paragraph above is good but feels repetitive to the first paragraph in the document and has minor problems. What information you are trying to deliver here? Can you summarize it here in the comments, please? And we'll see how can we improve the text.

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 20]
```diff
@@ -0,0 +1,148 @@
+# Shared Storage on NFS
+
+In the modern software development environment, teams are working together on
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: software development -> machine learning 

(I even agree that it's software engineering, but it's bette to delineate them for now)

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 21]
```diff
@@ -0,0 +1,148 @@
+# Shared Storage on NFS
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: I want to rename it ... it's not about NFS only. It's about any network attached storages. We can do something like:

`Share Storage on NAS (NFS)`

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+
+In the modern software development environment, teams are working together on
+same dataset to get the results. It became necessary that data is accessible and
+every team member has a same updated dataset. NFS (Network File System) storage
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: `NAS (NFS is one common example) is widely ... `

Comment (shcheklein) [reply static/docs/use-cases/shared-storage-on-nfs.md]: we can also mention something like: "Here we would like to show you how to setup a shared cache on NFS, but the same idea applies to any other NAS"

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+
+First configure NFS server and client machine, following this
+[link](https://vitux.com/install-nfs-server-and-client-on-ubuntu/).
+
+In order to make it work on a shared server, after configuring NFS server and
+client we need to setup a shared cache location for your projects, so that every
+team member is using the same cache location.
+
+After configuring NFS on both server and client side. Let's create an export
+directory on server side where all data will be stored.
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: it's better to use `:` when you have a code block you are writing about in the sentence

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+Let's create a mount point of client side.
+
+```dvc
+$ mkdir -p /mnt/dataset/
+```
+
+From `/mnt/dataset/` you will be able to access `/storage` directory present in
+host server from your local machine.
+
+## Configuring Cache location
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: location -> Location

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+$ git commit . -m "add first version of the dataset"
+$ git tag -a "v1.0" -m "dataset v1.0"
+$ git push origin HEAD
+$ git push origin v1.0
+```
+
+Next, you can easily get this appear in your workspace by:
+
+```dvc
+$ cd /home/user/project/
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: before the project path was `/project`

File: static/docs/use-cases/shared-storage-on-nfs.md [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+$ git pull
+$ dvc checkout
+```
+
+After `git pull`, you will be able to see a `data.dvc` file. To see more
+information on `.dvc` file format, visit
+[here](/doc/user-guide/dvc-file-format).
+
+`data` directory will now be a symbolic link to the NFS storage.
```
Comment (shcheklein) [inline static/docs/use-cases/shared-storage-on-nfs.md]: worth writing something similar to the last paragraph in the introduction to reiterate on why links are important, worth showing an output of the `ls -a`

**Comment by shcheklein** [comment]
@ryokugyu any updates on this? :) it's almost done as far as I can tell, would be great to get it merged.

**Comment by ryokugyu** [comment]
> @ryokugyu any updates on this? :) it's almost done as far as I can tell, would be great to get it merged.

@shcheklein will work on it. Sorry for the delay!

**Comment by dashohoxha** [comment]
I think that the "Mounted DVC Storage" (which is explained on this interactive example: https://katacoda.com/dvc/courses/examples/mounted-storage) is more general than just NFS and it deprecates this one.

**Comment by shcheklein** [comment]
> is more general than just NFS

my concern that it's very specific because of SSHFS and it's not emphasized enough that NFS, NAS (whatever else?) is covered

> deprecates this one

don't think so. Especially the way interactive tutorials are made - they are extremely dry and do not explain motivation very well, do not explain what is happening behind the scene and what commands are doing.

**Comment by dashohoxha** [comment]
> don't think so. Especially the way interactive tutorials are made - they are extremely dry and do not explain motivation very well, do not explain what is happening behind the scene and what commands are doing.

In this case it is just an interactive example (not a tutorial) and it is referenced from a User Guide page: https://dvc-org-pr-784.herokuapp.com/doc/user-guide/data-sharing/mounted-storage
So, the motivation and high level explanations are supposed to be elaborated on the UG page.

**Comment by shcheklein** [comment]
@dashohoxha 

> In this case it is just an interactive example (not a tutorial) and it is referenced from a User Guide page: https://dvc-org-pr-784.herokuapp.com/doc/user-guide/data-sharing/mounted-storage
So, the motivation and high level explanations are supposed to be elaborated on the UG page.

kk. It just in you initial comment you mentioned only the interactive tutorial and hadn't had enough time to see the UG changes. Will get back to this one when I have time to read the epic PR :)

**Comment by shcheklein** [comment]
@ryokugyu will keep it open for now. Still can be useful.

**Comment by shcheklein** [comment]
I think it is still relevant. Unfortunately, no easy way to reopen it now.
